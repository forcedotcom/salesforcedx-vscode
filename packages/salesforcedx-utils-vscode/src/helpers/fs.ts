/*
 * Copyright (c) 2025, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { dirname } from 'node:path';
import * as vscode from 'vscode';

export const readFile = async (filePath: string): Promise<string> => {
  try {
    const uri = vscode.Uri.file(filePath);
    const data = await vscode.workspace.fs.readFile(uri);
    return Buffer.from(data).toString('utf8');
  } catch (error) {
    throw new Error(`Failed to read file ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Writes content to a file
 * @param filePath The path to the file
 * @param content The content to write
 */
export const writeFile = async (filePath: string, content: string): Promise<void> => {
  try {
    const dirPath = dirname(filePath);
    if (!(await fileOrFolderExists(dirPath))) {
      await createDirectory(dirPath);
    }

    const encoder = new TextEncoder();
    const uint8Array = encoder.encode(content);
    await vscode.workspace.fs.writeFile(vscode.Uri.file(filePath), uint8Array);
  } catch (error) {
    throw new Error(`Failed to write file ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Checks if a file exists
 * @param filePath The path to the file
 * @returns True if the file exists, false otherwise
 */
export const fileOrFolderExists = async (filePath: string): Promise<boolean> => {
  try {
    const uri = vscode.Uri.file(filePath);
    await vscode.workspace.fs.stat(uri);
    return true;
  } catch {
    return false;
  }
};

/**
 * Creates a directory recursively.  Will not throw if the directory already exists.
 * @param dirPath The path to the directory
 */
export const createDirectory = async (dirPath: string): Promise<void> => {
  try {
    const uri = vscode.Uri.file(dirPath);
    await vscode.workspace.fs.createDirectory(uri);
  } catch (error) {
    throw new Error(`Failed to create directory ${dirPath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Deletes a file
 * @param filePath The path to the file
 */
export const deleteFile = async (
  filePath: string,
  options: { recursive?: boolean; useTrash?: boolean } = {}
): Promise<void> => {
  try {
    const uri = vscode.Uri.file(filePath);
    await vscode.workspace.fs.delete(uri, options);
  } catch (error) {
    throw new Error(`Failed to delete file ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Reads a directory's contents
 * @param dirPath The path to the directory
 * @returns Array of file/directory names
 */
export const readDirectory = async (dirPath: string): Promise<string[]> => {
  try {
    const uri = vscode.Uri.file(dirPath);
    const entries = await vscode.workspace.fs.readDirectory(uri);
    return entries.map(([name]) => name);
  } catch (error) {
    throw new Error(`Failed to read directory ${dirPath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Gets file stats
 * @param filePath The path to the file
 * @returns File stats including size, creation time, and modification time
 */
export const stat = async (filePath: string): Promise<vscode.FileStat> => {
  try {
    const uri = vscode.Uri.file(filePath);
    return await vscode.workspace.fs.stat(uri);
  } catch (error) {
    throw new Error(
      `Failed to get file stats for ${filePath}: ${error instanceof Error ? error.message : String(error)}`
    );
  }
};

export const safeDelete = async (
  filePath: string,
  options?: { recursive?: boolean; useTrash?: boolean }
): Promise<void> => {
  try {
    const uri = vscode.Uri.file(filePath);
    await vscode.workspace.fs.stat(uri);
    await vscode.workspace.fs.delete(uri, options);
  } catch {
    // File doesn't exist or can't be accessed, do nothing
  }
};

/**
 * Ensures the current working directory is set to the project path
 * @param rootWorkspacePath The path to the root workspace
 */
export const ensureCurrentWorkingDirIsProjectPath = async (rootWorkspacePath: string): Promise<void> => {
  if (rootWorkspacePath && process.cwd() !== rootWorkspacePath) {
    try {
      const uri = vscode.Uri.file(rootWorkspacePath);
      await vscode.workspace.fs.stat(uri);
      process.chdir(rootWorkspacePath);
    } catch {
      // Path doesn't exist, do nothing
    }
  }
};
