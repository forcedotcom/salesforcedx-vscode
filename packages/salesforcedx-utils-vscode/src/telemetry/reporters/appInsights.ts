/* eslint-disable header/header */
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/

'use strict';

import { TelemetryReporter } from '@salesforce/vscode-service-provider';
import * as appInsights from 'applicationinsights';
import * as os from 'os';
import { Disposable, env, UIKind, version, workspace } from 'vscode';
import { WorkspaceContextUtil } from '../../context/workspaceContextUtil';

type CommonProperties = {
  'common.os': string;
  'common.platformversion': string;
  'common.cpus'?: string;
  'common.systemmemory': string;
  'common.extname': string;
  'common.extversion': string;
  'common.vscodemachineid'?: string;
  'common.vscodesessionid'?: string;
  'common.vscodeversion'?: string;
  'common.vscodeuikind'?: string;
};

export class AppInsights extends Disposable implements TelemetryReporter {
  private appInsightsClient: appInsights.TelemetryClient | undefined;
  private userOptIn: boolean = false;
  private toDispose: Disposable[] = [];
  private uniqueUserMetrics: boolean = false;

  private static TELEMETRY_CONFIG_ID = 'telemetry';
  private static TELEMETRY_CONFIG_ENABLED_ID = 'enableTelemetry';

  // user defined tag to add to properties that is defined via setting
  private telemetryTag: string | undefined;

  constructor(
    private extensionId: string,
    private extensionVersion: string,
    key: string,
    readonly userId: string,
    enableUniqueMetrics?: boolean
  ) {
    super(() => this.toDispose.forEach(d => d && d.dispose()));
    if (enableUniqueMetrics) {
      this.uniqueUserMetrics = true;
    }
    this.setTelemetryTag();
    this.updateUserOptIn(key);
    this.toDispose.push(
      workspace.onDidChangeConfiguration(() => this.updateUserOptIn(key))
    );
  }

  private updateUserOptIn(key: string): void {
    const config = workspace.getConfiguration(AppInsights.TELEMETRY_CONFIG_ID);
    if (
      this.userOptIn !==
      config.get<boolean>(AppInsights.TELEMETRY_CONFIG_ENABLED_ID, true)
    ) {
      this.userOptIn = config.get<boolean>(
        AppInsights.TELEMETRY_CONFIG_ENABLED_ID,
        true
      );
      if (this.userOptIn) {
        this.createAppInsightsClient(key);
      } else {
        // tslint:disable-next-line:no-floating-promises
        this.dispose();
      }
    }
  }

  private createAppInsightsClient(key: string) {
    // check if another instance is already initialized
    if (appInsights.defaultClient) {
      this.appInsightsClient = new appInsights.TelemetryClient(key);
      // no other way to enable offline mode
      this.appInsightsClient.channel.setUseDiskRetryCaching(true);
    } else {
      appInsights
        .setup(key)
        .setAutoCollectRequests(false)
        .setAutoCollectPerformance(false)
        .setAutoCollectExceptions(false)
        .setAutoCollectDependencies(false)
        .setAutoDependencyCorrelation(false)
        .setAutoCollectConsole(false)
        .setUseDiskRetryCaching(true)
        .start();
      this.appInsightsClient = appInsights.defaultClient;
    }

    this.appInsightsClient.commonProperties = this.getCommonProperties();
    if (this.uniqueUserMetrics && env) {
      this.appInsightsClient.context.tags['ai.user.id'] = this.userId;
      this.appInsightsClient.context.tags['ai.session.id'] = env.sessionId;
      this.appInsightsClient.context.tags['ai.cloud.roleInstance'] =
        'DEPRECATED';
    }

    // check if it's an Asimov key to change the endpoint
    if (key && key.indexOf('AIF-') === 0) {
      this.appInsightsClient.config.endpointUrl =
        'https://vortex.data.microsoft.com/collect/v1';
    }
  }

  private getCommonProperties(): CommonProperties {
    const commonProperties: CommonProperties = {
      'common.os': os.platform(),
      'common.platformversion': (os.release() || '').replace(
        /^(\d+)(\.\d+)?(\.\d+)?(.*)/,
        '$1$2$3'
      ),
      'common.systemmemory': `${(
        os.totalmem() /
        (1024 * 1024 * 1024)
      ).toFixed(2)} GB`,
      'common.extname': this.extensionId,
      'common.extversion': this.extensionVersion
    };

    const cpus = os.cpus();
    if (cpus && cpus.length > 0) {
      commonProperties['common.cpus'] = `${cpus[0].model}(${cpus.length} x ${cpus[0].speed})`;
    }

    if (env) {
      commonProperties['common.vscodemachineid'] = env.machineId;
      commonProperties['common.vscodesessionid'] = env.sessionId;
      commonProperties['common.vscodeversion'] = version;
      if (env.uiKind) {
        commonProperties['common.vscodeuikind'] = UIKind[env.uiKind];
      }
    }

    return commonProperties;
  }

  public sendTelemetryEvent(
    eventName: string,
    properties?: { [key: string]: string },
    measurements?: { [key: string]: number }
  ): void {
    if (this.userOptIn && eventName && this.appInsightsClient) {
      const orgId = WorkspaceContextUtil.getInstance().orgId;
      let props = properties ? properties : {};
      props = this.applyTelemetryTag(orgId ? { ...props, orgId } : props);

      this.appInsightsClient.trackEvent({
        name: `${this.extensionId}/${eventName}`,
        // tslint:disable-next-line:object-literal-shorthand
        properties: props,
        // tslint:disable-next-line:object-literal-shorthand
        measurements
      });
    }
  }

  public sendExceptionEvent(
    exceptionName: string,
    exceptionMessage: string,
    measurements?: { [key: string]: number }
  ): void {
    if (this.userOptIn && exceptionMessage && this.appInsightsClient) {
      const error = new Error();
      error.name = `${this.extensionId}/${exceptionName}`;
      error.message = exceptionMessage;
      error.stack = 'DEPRECATED';

      const orgId = WorkspaceContextUtil.getInstance().orgId || '';
      const properties = this.applyTelemetryTag({ orgId });
      this.appInsightsClient.trackException({
        exception: error,
        properties,
        measurements
      });
    }
  }

  public dispose(): Promise<any> {
    const flushEventsToAI = new Promise<any>(resolve => {
      if (this.appInsightsClient) {
        this.appInsightsClient.flush({
          callback: () => {
            // all data flushed
            this.appInsightsClient = undefined;
            resolve(void 0);
          }
        });
      } else {
        resolve(void 0);
      }
    });
    return flushEventsToAI;
  }

  /**
   * Helper to set reporter's telemetryTag from setting salesforcedx-vscode-core.telemetry-tag
   * @returns string | undefined
   */
  private setTelemetryTag(): void {
    const config = workspace.getConfiguration();
    this.telemetryTag =
      config?.get('salesforcedx-vscode-core.telemetry-tag') || undefined;
  }

  /**
   * Helper to include telemetryTag in properties if it exists
   * if not, return properties as is
   *
   * @param properties
   * @returns
   */
  private applyTelemetryTag(properties: { [key: string]: string }): {
    [key: string]: string;
  } {
    return this.telemetryTag
      ? { ...properties, telemetryTag: this.telemetryTag }
      : properties;
  }
}
