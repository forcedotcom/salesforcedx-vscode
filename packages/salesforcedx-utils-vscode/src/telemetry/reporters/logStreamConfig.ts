/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
/**
 * Represents the configuration for a logging stream that is written to a local file.
 */
export class LogStreamConfig {
  /**
   * Retrieves the log file path.
   * @returns The log file path.
   */
  public static logFilePath(): string {
    const vsCodeLogsPath = process.env['VSCODE_LOGS'] ?? '';
    return vsCodeLogsPath;
  }

  /**
   * Checks if local stream logging is enabled for the specified extension.
   * @param extensionName - The name of the extension.
   * @returns True if logging is enabled, false otherwise.
   */
  public static isEnabledFor(extensionName: string): boolean | '' {
    const vsCodeLogLevelTrace = process.env['VSCODE_LOG_LEVEL'] === 'trace';
    return !!(
      LogStreamConfig.logFilePath() &&
      extensionName &&
      vsCodeLogLevelTrace
    );
  }
}
