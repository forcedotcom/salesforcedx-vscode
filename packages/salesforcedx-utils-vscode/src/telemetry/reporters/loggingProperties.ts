/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export type CommonProperties = {
  'common.os': string;
  'common.platformversion': string;
  'common.cpus'?: string;
  'common.systemmemory': string;
  'common.extname': string;
  'common.extversion': string;
  'common.vscodemachineid'?: string;
  'common.vscodesessionid'?: string;
  'common.vscodeversion'?: string;
  'common.vscodeuikind'?: string;
};

export type InternalProperties = {
  'sfInternal.hostname': string;
  'sfInternal.username': string;
};
