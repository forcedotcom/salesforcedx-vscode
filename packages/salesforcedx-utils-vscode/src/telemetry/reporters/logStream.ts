/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
'use strict';

import { TelemetryReporter } from '@salesforce/vscode-service-provider';
import * as fs from 'fs';
import * as path from 'path';
import { Disposable, workspace } from 'vscode';
import { WorkspaceContextUtil } from '../../context/workspaceContextUtil';

/**
 * Represents a telemetry reporter that logs telemetry events to a file.
 */
export class LogStream extends Disposable implements TelemetryReporter {
  private toDispose: Disposable[] = [];

  private stream: fs.WriteStream | undefined;

  constructor(
    private extensionId: string,
    private logFilePath: string
  ) {
    super(() => this.toDispose.forEach(d => d && d.dispose()));
    logFilePath = path.join(logFilePath, `${this.extensionId}.txt`);
    this.stream = fs.createWriteStream(logFilePath, {
      flags: 'a',
      encoding: 'utf8',
      autoClose: true
    });
    this.toDispose.push(workspace.onDidChangeConfiguration(() => () => {}));
    console.log(
      'VS Code telemetry event logging enabled for: ' +
        this.extensionId +
        '. Telemetry events will be written via write stream to a file at: ' +
        this.logFilePath +
        '.'
    );
  }

  public sendTelemetryEvent(
    eventName: string,
    properties?: { [key: string]: string },
    measurements?: { [key: string]: number }
  ): void {
    const orgId = WorkspaceContextUtil.getInstance().orgId;
    if (orgId && properties) {
      properties.orgId = orgId;
    } else if (orgId) {
      properties = { orgId };
    }

    if (this.stream) {
      this.stream.write(
        `telemetry/${eventName} ${JSON.stringify({
          properties,
          measurements
        })}\n`
      );
    }
  }

  public sendExceptionEvent(
    exceptionName: string,
    exceptionMessage: string,
    measurements?: { [key: string]: number }
  ): void {
    const orgId = WorkspaceContextUtil.getInstance().orgId || '';
    const properties = { orgId };
    console.log(
      'LogStream.sendExceptionEvent - exceptionMessage: ' + exceptionMessage
    );

    if (this.stream) {
      this.stream.write(
        `telemetry/${exceptionName} ${JSON.stringify({
          properties,
          measurements
        })}\n`
      );
    }
  }

  public dispose(): Promise<any> {
    const flushEventsToLogger = new Promise<any>(resolve => {
      if (!this.stream) {
        return resolve(void 0);
      }
      this.stream.on('finish', resolve);
      this.stream.end();
    });

    return flushEventsToLogger;
  }
}
