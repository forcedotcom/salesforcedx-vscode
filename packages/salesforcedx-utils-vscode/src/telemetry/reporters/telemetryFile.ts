/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as fs from 'fs';
import * as path from 'path';
import { getRootWorkspacePath } from '../..';
import { LOCAL_TELEMETRY_FILE } from '../../constants';
import { TelemetryReporter } from '../interfaces';

/**
 * Represents a telemetry file that logs telemetry events by appending to a local file.
 */
export class TelemetryFile implements TelemetryReporter {
  private filePath: string;

  constructor(extensionId: string) {
    this.filePath = this.logFilePathFor(extensionId);
    console.log(
      'Local telemetry event logging enabled for: ' +
        extensionId +
        '. Telemetry events will be appended to the file at: ' +
        this.filePath +
        '.'
    );
  }

  public sendTelemetryEvent(
    eventName: string,
    properties?: { [key: string]: string },
    measurements?: { [key: string]: number }
  ): void {
    void this.writeToFile(eventName, { ...properties, ...measurements });
  }

  public sendExceptionEvent(
    exceptionName: string,
    exceptionMessage: string,
    measurements?: { [key: string]: number }
  ): void {
    void this.writeToFile(exceptionName, { exceptionMessage, ...measurements });
  }

  public dispose(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        console.log('dispose called on Local Telemetry Logger.');
        resolve();
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * Writes telemetry data to a local file.
   * @param command - The command associated with the telemetry data.
   * @param data - The telemetry data to be written.
   */
  private async writeToFile(
    command: string,
    data: {
      [key: string]: string | number;
    }
  ) {
    const timestamp = new Date().toISOString();

    await fs.promises.appendFile(
      this.filePath,
      JSON.stringify({ timestamp, command, data }, null, 2) + ','
    );
  }

  private logFilePathFor(extensionId: string): string {
    return path.join(
      getRootWorkspacePath(),
      `${extensionId}-${LOCAL_TELEMETRY_FILE}`
    );
  }
}
