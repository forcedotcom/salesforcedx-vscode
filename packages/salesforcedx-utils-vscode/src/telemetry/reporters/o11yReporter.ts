/*
 * Copyright (c) 2025, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

'use strict';

import { TelemetryReporter } from '@salesforce/vscode-service-provider';
import * as os from 'node:os';
import { Disposable, env, UIKind, version, workspace } from 'vscode';
import { WorkspaceContextUtil } from '../../context/workspaceContextUtil';
import { O11yService } from '../../services/o11yService';
import { isInternalHost } from '../utils/isInternal';
import { CommonProperties, InternalProperties } from './loggingProperties';

export class O11yReporter extends Disposable implements TelemetryReporter {
  private userOptIn: boolean = false;
  private o11yUploadEndpoint: string;
  private toDispose: Disposable[] = [];
  private readonly o11yService: O11yService;

  // user defined tag to add to properties that is defined via setting
  private telemetryTag: string | undefined;

  constructor(
    private extensionId: string,
    private extensionVersion: string,
    o11yUploadEndpoint: string,
    readonly userId: string
  ) {
    super(() => this.toDispose.forEach(d => d && d.dispose()));
    this.o11yService = O11yService.getInstance();
    this.userOptIn = true; // Assume opt-in for now
    this.o11yUploadEndpoint = o11yUploadEndpoint;
    this.setTelemetryTag();
  }

  public async initialize(extensionName: string): Promise<void> {
    await this.o11yService.initialize(extensionName, this.o11yUploadEndpoint);
  }

  private getCommonProperties(): CommonProperties {
    const commonProperties: CommonProperties = {
      'common.os': os.platform(),
      'common.platformversion': (os.release() || '').replace(/^(\d+)(\.\d+)?(\.\d+)?(.*)/, '$1$2$3'),
      'common.systemmemory': `${(os.totalmem() / (1024 * 1024 * 1024)).toFixed(2)} GB`,
      'common.extname': this.extensionId,
      'common.extversion': this.extensionVersion
    };

    const cpus = os.cpus();
    if (cpus && cpus.length > 0) {
      commonProperties['common.cpus'] = `${cpus[0].model}(${cpus.length} x ${cpus[0].speed})`;
    }

    if (env) {
      commonProperties['common.vscodemachineid'] = env.machineId;
      commonProperties['common.vscodesessionid'] = env.sessionId;
      commonProperties['common.vscodeversion'] = version;
      if (env.uiKind) {
        commonProperties['common.vscodeuikind'] = UIKind[env.uiKind];
      }
    }

    return commonProperties;
  }

  private getInternalProperties(): InternalProperties {
    return {
      'sfInternal.hostname': os.hostname(),
      'sfInternal.username': os.userInfo().username
    };
  }

  private getUserProperties(): Record<string, string> {
    return {
      user_Id: this.userId,
      session_Id: env.sessionId
    };
  }

  private aggregateLoggingProperties() {
    const commonProperties = { ...this.getUserProperties(), ...this.getCommonProperties() };
    return isInternalHost() ? { ...commonProperties, ...this.getInternalProperties() } : commonProperties;
  }

  public sendTelemetryEvent(
    eventName: string,
    properties?: { [key: string]: string },
    measurements?: { [key: string]: number }
  ): void {
    if (this.userOptIn && eventName) {
      const orgId = WorkspaceContextUtil.getInstance().orgId;
      const orgShape = WorkspaceContextUtil.getInstance().orgShape || '';
      const devHubId = WorkspaceContextUtil.getInstance().devHubId || '';
      let props = properties ? { ...properties, ...this.aggregateLoggingProperties() } : {};
      props = this.applyTelemetryTag(orgId ? { ...props, orgId, orgShape, devHubId } : props);

      this.o11yService.logEvent({
        name: `${this.extensionId}/${eventName}`,
        properties: props,
        measurements
      });
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      this.o11yService.upload();
    }
  }

  public sendExceptionEvent(
    exceptionName: string,
    exceptionMessage: string,
    measurements?: { [key: string]: number }
  ): void {
    if (this.userOptIn && exceptionMessage) {
      const error = new Error();
      error.name = `${this.extensionId}/${exceptionName}`;
      error.message = exceptionMessage;
      error.stack = 'DEPRECATED';

      const orgId = WorkspaceContextUtil.getInstance().orgId || '';
      const orgShape = WorkspaceContextUtil.getInstance().orgShape || '';
      const devHubId = WorkspaceContextUtil.getInstance().devHubId || '';
      const properties = this.applyTelemetryTag({ orgId, orgShape, devHubId });
      this.o11yService.logEvent({
        exception: error,
        properties,
        measurements
      });

      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      this.o11yService.upload();
    }
  }

  public async dispose(): Promise<void> {
    await this.o11yService.upload();
  }

  /**
   * Helper to set reporter's telemetryTag from setting salesforcedx-vscode-core.telemetry-tag
   * @returns string | undefined
   */
  private setTelemetryTag(): void {
    const config = workspace.getConfiguration();
    this.telemetryTag = config?.get('salesforcedx-vscode-core.telemetry-tag') || undefined;
  }

  /**
   * Helper to include telemetryTag in properties if it exists
   * if not, return properties as is
   *
   * @param properties
   * @returns
   */
  private applyTelemetryTag(properties: { [key: string]: string }): {
    [key: string]: string;
  } {
    return this.telemetryTag ? { ...properties, telemetryTag: this.telemetryTag } : properties;
  }
}
