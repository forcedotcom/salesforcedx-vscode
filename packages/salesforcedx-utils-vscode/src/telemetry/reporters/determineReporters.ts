/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { TelemetryReporter } from '@salesforce/vscode-service-provider';
import { isLocalLogging } from '../../telemetry/utils/devModeUtils';
import { AppInsights } from './appInsights';
import { LogStream } from './logStream';
import { LogStreamConfig } from './logStreamConfig';
import { TelemetryFile } from './telemetryFile';
import { TelemetryReporterConfig } from './telemetryReporterConfig';

export const determineReporters = (config: TelemetryReporterConfig) => {
  const {extName, version, aiKey, userId, reporterName, isDevMode} = config;
  const reporters: TelemetryReporter[] = [];

  if(isDevMode) {
    addDevModeReporter(reporters, extName);
  } else {
    addAppInsightsReporter(reporters, reporterName, version, aiKey, userId);
    addLogstreamReporter(reporters, extName);
  }
  return reporters;
};

const addDevModeReporter = (reporters: TelemetryReporter[], extName: string) => {
  if(isLocalLogging(extName)) {
    // The new TelemetryFile reporter is run in Dev mode, and only
    // requires the advanced setting to be set re: configuration.
    reporters.push(new TelemetryFile(extName));
  }
};

const addAppInsightsReporter = (reporters: TelemetryReporter[], reporterName: string, version: string, aiKey: string, userId: string) => {
  console.log('adding AppInsights reporter.');
  reporters.push(
    new AppInsights(
      reporterName,
      version,
      aiKey,
      userId,
      true
    )
  );
};

/*
* Assuming this fs streaming is more efficient than the appendFile technique that
* the new TelemetryFile reporter uses, I am keeping the logic in place for which
* reporter is used when.  The original log stream functionality only worked under
* the same conditions as the AppInsights capabilities, but with additional configuration.
*/
const addLogstreamReporter = (reporters: TelemetryReporter[], extName: string) => {
  if(LogStreamConfig.isEnabledFor(extName)) {
    reporters.push(
      new LogStream(
        extName,
        LogStreamConfig.logFilePath()
      )
    );
  }
};
