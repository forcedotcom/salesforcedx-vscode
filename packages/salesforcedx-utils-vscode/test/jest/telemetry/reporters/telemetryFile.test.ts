/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as fs from 'fs';
import { TelemetryFile } from '../../../../src/telemetry/reporters/telemetryFile';

jest.mock('fs', () => ({
  ...jest.requireActual('fs'),
  promises: {
    appendFile: jest.fn().mockResolvedValue({})
  }
}));

describe('TelemetryFile', () => {
  let telemetryFile: TelemetryFile;
  let writeToFileMock: jest.SpyInstance;
  const dummyExtensionId = 'extensionId';

  beforeEach(() => {
    telemetryFile = new TelemetryFile(dummyExtensionId);
    writeToFileMock = jest.spyOn(telemetryFile as any, 'writeToFile');
  });

  it('should append the telemetry event to the telemetry file.', () => {
    const eventName = 'testEvent';
    const properties = { key1: 'value1', key2: 'value2' };
    const measurements = { metric1: 10, metric2: 20 };

    telemetryFile.sendTelemetryEvent(eventName, properties, measurements);

    expect(writeToFileMock).toHaveBeenCalledWith(eventName, {
      ...properties,
      ...measurements
    });
  });

  it('should append the exception event to the telemetry file', () => {
    const exceptionName = 'testException';
    const exceptionMessage = 'Test exception message';
    const measurements = { metric1: 10, metric2: 20 };

    telemetryFile.sendExceptionEvent(
      exceptionName,
      exceptionMessage,
      measurements
    );

    expect(writeToFileMock).toHaveBeenCalledWith(exceptionName, {
      exceptionMessage,
      ...measurements
    });
  });

  it('should dispose', async () => {
    await expect(telemetryFile.dispose()).resolves.toBeUndefined();
  });

  describe('writeToFile', () => {
    it('should append data to the telemetry file', async () => {
      jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));
      const eventName = 'testEvent';
      const properties = { key1: 'value1', key2: 'value2' };

      await (telemetryFile as any).writeToFile(eventName, properties);

      expect(
        (fs.promises.appendFile as jest.Mock).mock.calls[0]
      ).toMatchSnapshot();
    });
  });
});
