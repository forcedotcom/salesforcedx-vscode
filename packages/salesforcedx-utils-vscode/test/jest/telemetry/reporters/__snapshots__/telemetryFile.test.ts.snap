// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TelemetryFile writeToFile should append data to the telemetry file 1`] = `
[
  "extensionId-telemetry.json",
  "{
  "timestamp": "2020-01-01T00:00:00.000Z",
  "command": "testEvent",
  "data": {
    "key1": "value1",
    "key2": "value2"
  }
},",
]
`;
